ocaml/libs/lru-cache/lru.ml

```diff
@@ -0,0 +1,294 @@
+(*
+ * This module implements a cache with support for a least-recently-used
+ * (LRU) replacement policy. Main features:
+ *
+ * Implemented with standard OCaml data types, no dependency on outside
+ * libararies for the main functionality.
+ *
+ * Implemented with mutable state. This keeps the implementation compact
+ * and efficient but requires thinking about state more.
+ *
+ * The architecture is: elements are kept in a hash table to look them
+ * up based on a key. Additionally they are kept in a doubly linked
+ * list. The head of the list is the least recently used element that
+ * can be dropped to make room in the cache. When an element is found in
+ * the cache it is moved to the tail of the linked list.
+ *)
+
+let invalid_arg fmt = Printf.kprintf invalid_arg fmt
+
+let _fail fmt = Printf.kprintf failwith fmt
+
+module LL : sig
+  (** Doubly linked list ['a t] holding elements of type ['a]. *)
+
+  (** doubly linked list; this is a cyclic data structure; don't use [=]
```

Comment:
It *can* be implemented as a cyclic data structure, but the current one doesn't *seem* to be: previous of first is None and next of last is None. Probably best to avoid using '=' anyway though.

Reply:
This is cyclic: you have a cycle between neighbouring elements but not between first/last. The cycle is the reason that `=` must not be used.